pkg/iptables/iptables.go

```diff
@@ -58,81 +111,88 @@ func NewTrafficRedirector(
 	}, nil
 }
 
-// delete iptables rules for redirection
-func (tr *redirector) deleteRedirectRules() error {
-	return tr.execRedirectCmd("-D")
-}
-
-// add iptables rules for redirection
-func (tr *redirector) addRedirectRules() error {
-	return tr.execRedirectCmd("-A")
-}
-
-// add iptables rules for reset connections to port
-func (tr *redirector) addResetRules(port uint) error {
-	return tr.execResetCmd("-A", port)
+func (tr *redirector) redirectRules() []string {
+	return []string{
+		fmt.Sprintf(
+			redirectLocalRule,
+			tr.TargetPort,
+			tr.ProxyPort,
+		),
+		fmt.Sprintf(
+			redirectExternalRule,
+			tr.TargetPort,
+			tr.ProxyPort,
+		),
+	}
 }
 
-// delete iptables rules for reset connections to port
-func (tr *redirector) deleteResetRules(port uint) error {
-	return tr.execResetCmd("-D", port)
+func (tr *redirector) resetRules() []string {
+	return []string{
+		fmt.Sprintf(
+			resetLocalRule,
+			tr.TargetPort,
+		),
+		fmt.Sprintf(
+			resetExternalRule,
+			tr.TargetPort,
+		),
+	}
 }
 
-// buildRedirectCmd builds a command for adding or removing a transparent proxy using iptables
-func (tr *redirector) execRedirectCmd(action string) error {
-	cmd := fmt.Sprintf(
-		redirectCommand,
-		action,
-		tr.Iface,
-		tr.DestinationPort,
-		tr.RedirectPort,
-	)
-
+// execIptables runs performs the specified action ("-A" or "-D") for the supplied rule.
+func (tr *redirector) execIptables(action string, rule string) error {
+	cmd := fmt.Sprintf("%s %s", action, rule)
 	out, err := tr.executor.Exec("iptables", strings.Split(cmd, " ")...)
 	if err != nil {
-		return fmt.Errorf("error executing iptables command: %w %s", err, string(out))
+		return fmt.Errorf("error executing iptables command %q: %w %s", cmd, err, string(out))
 	}
 
 	return nil
 }
 
-func (tr *redirector) execResetCmd(action string, port uint) error {
-	cmd := fmt.Sprintf(
-		resetCommand,
-		action,
-		tr.Iface,
-		port,
-	)
+// Start applies the TrafficRedirect
+func (tr *redirector) Start() error {
+	for _, rule := range tr.redirectRules() {
+		err := tr.execIptables("-A", rule)
+		if err != nil {
+			return err
+		}
+	}
 
-	out, err := tr.executor.Exec("iptables", strings.Split(cmd, " ")...)
-	if err != nil {
-		return fmt.Errorf("error executing iptables command: %s", string(out))
+	for _, rule := range tr.resetRules() {
+		err := tr.execIptables("-A", rule)
+		if err != nil {
+			return err
+		}
 	}
 
 	return nil
 }
 
-// Starts applies the TrafficRedirect
-func (tr *redirector) Start() error {
-	// error is ignored as the rule may not exist
-	_ = tr.deleteResetRules(tr.RedirectPort)
-	if err := tr.addRedirectRules(); err != nil {
-		return err
+// Stop stops the TrafficRedirect.
+// Stop will continue attempting to remove all the rules it deployed even if removing one fails.
+// TODO: The error returned does not wrap original errors.
+func (tr *redirector) Stop() error {
```

Comment:
A small behavior change is that `Stop` no longer leaves a rule rejecting traffic to the proxy port. This can be added back if we deem it useful.

Reply:
This is not hard to add at all, but I wonder if it is something we _need_. I do not particularly like leaving iptables rules after disruption finishes, as those would stick there potentially forever without the user really knowing it.

Terminating the proxy server should already be killing existing connections, right? Happy to hear your thoughts about this @pablochacin

Reply:
I'm not sure about the implications. Maybe there are none, but we could possibly compare a TCP dump of both scenarios while running a test that continues after the proxy ends, and see if there's any significant difference.  But I don't think this should be a concern for this PR.

Reply:
I've added the rule back with a TODO for us to revisit this 👍🏻